vmap support for linalg.lu_factor

isdanni opened the pull request:
Differential Revision: D43093457

Fix #91415

### Expected behaviour

No use warning. 

```python
from functorch import vmap
x = torch.randn(4, 3, 2)
z = vmap(torch.linalg.lu_factor)(x)
```
Same behaviour as for-loop:

```python
x = torch.randn(4, 3, 2)
results = []
for xi in x:
  y = torch.linalg.lu_factor(xi)
  results.append(y)
```

Skylion007 commented on aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:
```diff
@@ -337,12 +337,20 @@ oneOutput cholesky_solve_batch_rule(
+    const Tensor& A, c10::optional<int64_t> A_bdim, bool pivot) {
+  TORCH_CHECK(rankWithoutBatchDim(A, A_bdim) >= 2, "torch.lu_factor: Expected tensor with 2 or more dimensions. Got size: ", A.sizes(), " instead");
+  const auto A_ = moveBatchDimToFront(A, A_bdim);
+  const auto res = at::linalg_lu_factor(A_, pivot);
```
Really tiny nit, but it might better to use structure bindings here now that we are on C++17. Would make it a bit more readable.

isdanni committed:
vmap support for linalg.lu_factor (#94328)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/94328

Test Plan: Please refer to GitHub actions

Reviewed By: zou3519, Chillee

Differential Revision: D43093457

fbshipit-source-id: 93d760c15ab830cc92d57108841a73e51b9b1a0e
